exporter/awsemfexporter/grouped_metric.go

```diff
@@ -224,3 +177,147 @@ func translateUnit(metric pmetric.Metric, descriptor map[string]MetricDescriptor
 	}
 	return unit
 }
+
+func shouldCompactMetrics(pmd pmetric.Metric, config *Config) bool {
+	if pmd.Type() != pmetric.MetricTypeGauge {
+		return false
+	}
+	// Check if the current metric is in the GaugeMetricsToCompact list
+	for _, name := range config.GaugeMetricsToCompact {
+		if name == pmd.Name() {
+			return true
+		}
+	}
+	return false
+}
+
+func filterAndCalculateDps(dps dataPoints, metricName string, metadata cWMetricMetadata, config *Config, calculators *emfCalculators) []dataPoint {
+	var result []dataPoint
+	for i := 0; i < dps.Len(); i++ {
+		if isStale, attrs := dps.IsStaleNaNInf(i); isStale {
+			if config != nil && config.logger != nil {
+				config.logger.Debug("dropped metric with nan value",
+					zap.String("metric.name", metricName),
+					zap.Any("metric.attributes", attrs))
+			}
+			continue
+		}
+		calculated, retained := dps.CalculateDeltaDatapoints(i, metadata.instrumentationScopeName, config.DetailedMetrics, calculators)
+		if retained {
+			result = append(result, calculated...)
+		}
+	}
+	return result
+}
+
+func enrichLabels(dp dataPoint, config *Config) map[string]string {
+	labels := dp.labels
+	if metricType, ok := labels["Type"]; ok {
+		if (metricType == "Pod" || metricType == "Container") && config.EKSFargateContainerInsightsEnabled {
+			addKubernetesWrapper(labels)
+		}
+	}
+	return labels
+}
+
+func replacePatternsIfNeeded(metadata cWMetricMetadata, labels map[string]string, config *Config, patternReplaceSucceeded bool) cWMetricMetadata {
+	if !patternReplaceSucceeded {
+		if strings.Contains(metadata.logGroup, "undefined") {
+			metadata.logGroup, _ = replacePatterns(config.LogGroupName, labels, config.logger)
+		}
+		if strings.Contains(metadata.logStream, "undefined") {
+			metadata.logStream, _ = replacePatterns(config.LogStreamName, labels, config.logger)
+		}
+	}
+	return metadata
+}
+
+func upsertGroupedMetric(
+	groupedMetrics map[any]*groupedMetric,
+	metadata cWMetricMetadata,
+	labels map[string]string,
+	metricName string,
+	metricVal any,
+	unit string,
+	logger *zap.Logger,
+) {
+	metric := &metricInfo{value: metricVal, unit: unit}
+	groupKey := aws.NewKey(metadata.groupedMetricMetadata, labels)
+
+	if _, ok := groupedMetrics[groupKey]; ok {
+		// if MetricName already exists in metrics map, print warning log
+		if _, ok := groupedMetrics[groupKey].metrics[metricName]; ok {
+			logger.Warn("Duplicate metric found", zap.String("Name", metricName), zap.Any("Labels", labels))
+		} else {
+			groupedMetrics[groupKey].metrics[metricName] = metric
+		}
+	} else {
+		groupedMetrics[groupKey] = &groupedMetric{
+			labels:   labels,
+			metrics:  map[string]*metricInfo{metricName: metric},
+			metadata: metadata,
+		}
+	}
+}
+
+// compactGaugeMetrics converts a collection of gauge data points into a compact representation, e.g. values and counts.
+func compactGaugeMetrics(
```

Comment:
iiuc it's not really compacting gauges but converting to CWHistogram. The name could be more aligned with what it actually does like `convertGaugesToCWHistogram` or something